File: mobile/src/main/java/org/dwallach/calwatch/PhoneActivity.java
/*
 * CalWatch
 * Copyright (C) 2014 by Dan Wallach
 * Home page: http://www.cs.rice.edu/~dwallach/calwatch/
 * Licensing: http://www.cs.rice.edu/~dwallach/calwatch/licensing.html
 */
package org.dwallach.calwatch;
import android.app.Activity;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.RadioButton;
import android.widget.Switch;

import java.lang.ref.WeakReference;
import java.util.Observable;
import java.util.Observer;


public class PhoneActivity extends Activity implements Observer {
    private final static String TAG = "PhoneActivity";

    private RadioButton toolButton, numbersButton, liteButton;
    private MyViewAnim clockView;
    private Switch secondsSwitch, dayDateSwitch;

    private ClockState clockState;
    private boolean disableUICallbacks = false;

    private static WeakReference<Activity> activityRef;

    private ClockState getClockState() {
        if(clockState == null) {
            Log.v(TAG, "reconnecting clock state");
            clockState = ClockState.getSingleton();
            clockState.addObserver(this);
        }
        return clockState;
    }

    public PhoneActivity() {
        super();
        activityRef = new WeakReference<Activity>(this);
    }

    //
    // this will be called, eventually, from whatever feature is responsible for
    // restoring saved user preferences
    //
    private void setFaceModeUI(int mode, boolean showSeconds, boolean showDayDate) {
        Log.v(TAG, "setFaceModeUI");
        if(toolButton == null || numbersButton == null || liteButton == null || secondsSwitch == null || dayDateSwitch == null) {
            Log.v(TAG, "trying to set UI mode without buttons active yet");
            return;
        }

        disableUICallbacks = true;

        try {
            switch (mode) {
                case ClockState.FACE_TOOL:
                    toolButton.performClick();
                    break;
                case ClockState.FACE_NUMBERS:
                    numbersButton.performClick();
                    break;
                case ClockState.FACE_LITE:
                    liteButton.performClick();
                    break;
                default:
                    Log.v(TAG, "bogus face mode: " + mode);
                    break;
            }

            secondsSwitch.setChecked(showSeconds);
            dayDateSwitch.setChecked(showDayDate);
        } catch(Throwable throwable) {
            // probably a called-from-wrong-thread-exception, we'll just ignore it
            Log.v(TAG, "ignoring exception while updating button state");
        }

        disableUICallbacks = false;
    }

    private void getFaceModeFromUI() {
        Log.v(TAG, "getFaceModeFromUI");
        int mode = -1;

        if(toolButton == null || numbersButton == null || liteButton == null || secondsSwitch == null || dayDateSwitch == null) {
            Log.v(TAG, "trying to get UI mode without buttons active yet");
            return;
        }

        if(toolButton.isChecked())
            mode = ClockState.FACE_TOOL;
        else if(numbersButton.isChecked())
            mode = ClockState.FACE_NUMBERS;
        else if(liteButton.isChecked())
            mode = ClockState.FACE_LITE;
        else Log.v(TAG, "no buttons are selected? weird.");

        boolean showSeconds = secondsSwitch.isChecked();
        boolean showDayDate = dayDateSwitch.isChecked();

        if(mode != -1) {
            getClockState().setFaceMode(mode);
        }
        getClockState().setShowSeconds(showSeconds);
        getClockState().setShowDayDate(showDayDate);

        getClockState().pingObservers(); // we only need to do this once, versus a whole bunch of times when it was happening internally
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Log.v(TAG, "Create!");

        setContentView(R.layout.activity_phone);
    }

    @Override
    protected void onPause() {
        super.onPause();
        Log.v(TAG, "Pause!");

        // perhaps incorrect assumption: if our activity is being killed, onStop will happen beforehand,
        // so we'll deregister our clockState observer, allowing this Activity object to become
        // garbage. A new one will be created if the activity ever comes back to life, which
        // will call getClockState(), which will in turn resurrect observer. Setting clockState=null
        // means that, even if this specific Activity object is resurrected from the dead, we'll
        // just reconnect it the next time somebody internally calls getClockState(). No harm, no foul.

        // http://developer.android.com/reference/android/app/Activity.html

        if (clockView != null)
            clockView.kill(this);

        if(clockState != null) {
            clockState.deleteObserver(this);
            clockState = null;
        }
    }

    @Override
    protected void onStart() {
        super.onStart();
        Log.v(TAG, "Start!");

        // Core UI widgets: find 'em
        liteButton = (RadioButton) findViewById(R.id.liteButton);
        toolButton = (RadioButton) findViewById(R.id.toolButton);
        numbersButton = (RadioButton) findViewById(R.id.numbersButton);
        clockView = (MyViewAnim) findViewById(R.id.surfaceView);
        secondsSwitch = (Switch) findViewById(R.id.showSeconds);
        dayDateSwitch = (Switch) findViewById(R.id.showDayDate);
//        clockView.setSleepInEventLoop(true);

        Log.v(TAG, "registering callback");

        View.OnClickListener myListener = new View.OnClickListener() {
            public void onClick(View v) {
                if(!disableUICallbacks)
                    getFaceModeFromUI();
                if(clockView != null)
                    clockView.invalidate();
            }
        };

        liteButton.setOnClickListener(myListener);
        toolButton.setOnClickListener(myListener);
        numbersButton.setOnClickListener(myListener);
        secondsSwitch.setOnClickListener(myListener);
        dayDateSwitch.setOnClickListener(myListener);

        WatchCalendarService.kickStart(this);  // bring it up, if it's not already up
        PreferencesHelper.loadPreferences(this);
        CalendarPermission.init(this);

        clockView.init(this);
        clockView.initCalendarFetcher(this);

        onResume();
        Log.v(TAG, "activity setup complete");
    }

    @Override
    protected void onResume() {
        super.onResume();
        Log.v(TAG, "Resume!");

        getClockState(); // side-effects: re-initializes observer

        if(clockView != null) {
            clockView.init(this);
            clockView.invalidate();
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
//        getMenuInflater().inflate(R.menu.phone, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void update(Observable observable, Object data) {
        // somebody changed *something* in the ClockState, causing us to get called
        Log.v(TAG, "Noticed a change in the clock state; saving preferences");
        setFaceModeUI(getClockState().getFaceMode(), getClockState().getShowSeconds(), getClockState().getShowDayDate());
        PreferencesHelper.savePreferences(this);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String permissions[], @NonNull int[] grantResults) {
        Log.v(TAG, "onRequestPermissionsResult");
        CalendarPermission.handleResult(requestCode, permissions, grantResults);
//        CalWatchFaceService.Engine engine = CalWatchFaceService.getEngine();
//        if(engine != null) {
//            engine.calendarPermissionUpdate();
//        }
        Log.v(TAG, "finishing PermissionActivity");
    }

    /**
     * This will be called when the user clicks on the watchface, presumably because they want
     * us to request calendar permissions.
     */
    static void watchfaceClick(MyViewAnim view) {
        Log.v(TAG, "Watchface clicked!");
        if(activityRef == null)
            return; // can't do anything without an activity

        Activity activity = activityRef.get();
        if(activity == null)
            return; // can't do anything with an activity

        ClockState clockState = ClockState.getSingleton();
        if(clockState == null)
            Log.v(TAG, "Activity found and clockState is null.");
        else if(!clockState.getCalendarPermission()) {
            Log.v(TAG, "Requesting permissions");
            CalendarPermission.request(activity);
            view.initCalendarFetcher(activity);
        } else {
            Log.v(TAG, "Permissions already granted.");
        }
    }
}

File: mobile/src/main/kotlin/org/dwallach/calwatch/PhoneActivity.kt
/*
 * CalWatch
 * Copyright (C) 2014 by Dan Wallach
 * Home page: http://www.cs.rice.edu/~dwallach/calwatch/
 * Licensing: http://www.cs.rice.edu/~dwallach/calwatch/licensing.html
 */
package org.dwallach.calwatch

import android.app.Activity
import android.content.ContextWrapper
import android.graphics.drawable.AnimationDrawable
import android.os.Build
import android.os.Bundle
import android.support.v4.content.ContextCompat
import android.support.v7.widget.CardView
import android.view.Menu
import android.view.MenuItem
import android.view.View

import java.util.Observable
import java.util.Observer

import kotlinx.android.synthetic.main.activity_phone_intl.*
import org.jetbrains.anko.*

class PhoneActivity : Activity(), Observer, AnkoLogger {
    private var disableUICallbacks = false

    private fun select(card: CardView, button: View, selected: Boolean) {
        val bgColor = ContextCompat.getColor(this, if (selected) R.color.calWatchDark else R.color.calWatchPrimary)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            // TODO: do something useful with the button, not just the card -- tinting?
            card.setCardBackgroundColor(bgColor)
        } else {
            // On Android 4.4, setting the Card background color does nothing, and there's no paint filling either,
            // so we'll fall back to monkeying around with the button itself
            button.backgroundColor = bgColor
        }
    }

    /**
     * This method loads the UI state with the given mode (ClockState.FACE_NUMBERS, etc.) and other
     * features. Call this whenever necessary.
     */
    private fun setFaceModeUI(mode: Int, showSecondsP: Boolean, showDayDateP: Boolean, showStepCounterP: Boolean) {
        verbose { "setFaceModeUI: mode($mode), showSecondsP($showSecondsP), showDayDateP($showDayDateP), showStepCounterP($showStepCounterP)" }
        if (!uiButtonsReady()) {
            verbose("trying to set UI mode without buttons active yet")
            return
        }

        disableUICallbacks = true

        try {
            // while we're here, we'll also put the proper day/date into the relevant button
            val dayOfWeek = TimeWrapper.localDayOfWeek()
            val monthDay = TimeWrapper.localMonthDay()
            dayDateButton.text = "$monthDay\n$dayOfWeek"

            when(ClockState.faceMode) {
                ClockState.FACE_NUMBERS -> {
                    select(toolCard, toolButton, false)
                    select(numbersCard, numbersButton, true)
                    select(liteCard, liteButton, false)
                }
                ClockState.FACE_LITE -> {
                    select(toolCard, toolButton, false)
                    select(numbersCard, numbersButton, false)
                    select(liteCard, liteButton, true)
                }
                ClockState.FACE_TOOL -> {
                    select(toolCard, toolButton, true)
                    select(numbersCard, numbersButton, false)
                    select(liteCard, liteButton, false)
                }
            }

            select(secondsImageCard, secondsImageButton, ClockState.showSeconds)
            select(stepCountImageCard, stepCountImageButton, ClockState.showStepCounter)
            select(dayDateCard, dayDateButton, ClockState.showDayDate)

        } catch (throwable: Throwable) {
            // probably a called-from-wrong-thread-exception, we'll just ignore it
            verbose("ignoring exception while updating button state")
        }

        disableUICallbacks = false
    }

    private fun uiButtonsReady() :Boolean =
            toolButton != null && numbersButton != null && liteButton != null


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        verbose("Create!")

        setContentView(R.layout.activity_phone_intl)
    }

    override fun onWindowFocusChanged(hasFocus: Boolean) {
        //
        // We're using an animated button for letting the user select the seconds hand; the blurb
        // below is necessary to start the animation.
        //
        // http://developer.android.com/guide/topics/graphics/drawable-animation.html
        //

        val secondsAnim = secondsImageButton.drawable as AnimationDrawable

        if(hasFocus)
            secondsAnim.start()
        else
            secondsAnim.stop()
    }

    override fun onPause() {
        super.onPause()
        verbose("Pause!")

        surfaceView.pause(this)

        ClockState.deleteObserver(this)
    }

    private fun setupClickListener(clickMe: View?, logMe: String, callMe: () -> Unit) {
        clickMe?.setOnClickListener {
            verbose(logMe)
            callMe()

            ClockState.pingObservers() // side-effect: will call back to PhoneActivity to update the UI and save state
            surfaceView.invalidate() // redraws the clock
        }
    }

    override fun onStart() {
        super.onStart()
        verbose("Start!")

        if(!uiButtonsReady()) {
            error("UI buttons aren't ready yet")
        } else {
            setupClickListener(liteButton, "lite-mode selected") {
                ClockState.faceMode = ClockState.FACE_LITE
            }

            setupClickListener(toolButton, "tool-mode selected") {
                ClockState.faceMode = ClockState.FACE_TOOL
            }

            setupClickListener(numbersButton, "numbers-mode selected") {
                ClockState.faceMode = ClockState.FACE_NUMBERS
            }

            // Each of these handles the big image buttons that we want to also cause
            // their switches to toggle.

            setupClickListener(secondsImageButton, "seconds toggle") {
                ClockState.showSeconds = !ClockState.showSeconds
            }

            setupClickListener(dayDateButton, "dayDate toggle") {
                ClockState.showDayDate = !ClockState.showDayDate
            }

            setupClickListener(stepCountImageButton, "stepCount toggle") {
                ClockState.showStepCounter = !ClockState.showStepCounter
            }
        }

        WatchCalendarService.kickStart(this)  // bring it up, if it's not already up
        PreferencesHelper.loadPreferences(this)
        CalendarPermission.init(this)
        GoogleApiWrapper.startConnection(this) { verbose { "GoogleApi ready" } }

        surfaceView.init(this)
        surfaceView.initCalendarFetcher(this)

        onResume()
        verbose("activity setup complete")
    }

    override fun onResume() {
        super.onResume()
        verbose("Resume!")

        GoogleApiWrapper.startConnection(this) { verbose { "GoogleApi ready" } }
        ClockState.addObserver(this)
        surfaceView.resume(this)
        surfaceView.invalidate()
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        //        getMenuInflater().inflate(R.menu.phone, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        return super.onOptionsItemSelected(item)
    }

    override fun update(observable: Observable?, data: Any?) {
        verbose("Noticed a change in the clock state; saving preferences")

        setFaceModeUI(ClockState.faceMode, ClockState.showSeconds, ClockState.showDayDate, ClockState.showStepCounter)
        PreferencesHelper.savePreferences(this)
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        verbose("onRequestPermissionsResult")

        CalendarPermission.handleResult(requestCode, permissions, grantResults)
        surfaceView.initCalendarFetcher(this)

        verbose("finishing PermissionActivity")
    }

    companion object: AnkoLogger {
        /**
         * This will be called when the user clicks on the watchface, presumably because they want
         * us to request calendar permissions.
         */
        fun watchfaceClick(view: MyViewAnim) {
            verbose("Watchface clicked!")

            val activity = view.toActivity()

            if (!ClockState.calendarPermission) {
                verbose("Requesting permissions")
                CalendarPermission.request(activity)
                view.initCalendarFetcher(activity)
            } else {
                verbose("Permissions already granted.")
            }
        }
    }
}

/**
 * Helper extension function to convert from a view to its surrounding activity, if that activity exists.
 */
fun View.toActivity(): Activity {
    // See: http://stackoverflow.com/questions/8276634/android-get-hosting-activity-from-a-view
    var context = this.context
    while(context is ContextWrapper) {
        if(context is Activity) {
            return context
        } else {
            context = context.baseContext
        }
    }
    throw NoSuchFieldError("no activity found for this view") // shouldn't ever happen
}
